### ISSUE TITLE
[no-unnecessary-condition] Undefined is not considered to be a condition

### ISSUE THREAD

**Issue by MiK546** [typescript-eslint/typescript-eslint#1829]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": [
      "error",
      {
        "checkArrayPredicates": true
      }
    ]
  }
}
```

```TS
function testFunction(param: "string1" | "string2" | undefined){
  // If param is not defined, don't do anything
  if(param === "string1"){
    doSomething();
  }else if(param === "string2"){
    doSomethingElse();
  }
}
```

**Expected Result**
This example code should not produce a linting error with the above rule as the `param` parameter could be undefined and thus we need to check for the value of `param` in both conditionals.

**Actual Result**
The `else if` clause produces a linting error:
```
error  Unnecessary conditional, both sides of the expression are literal values  @typescript-eslint/no-unnecessary-condition
```
even though it is necessary to make sure `param` is not undefined.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.25.0` |
| `@typescript-eslint/parser`        | `2.25.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.1` |
| `npm`                              | `6.13.4` |

**Comment by bradzacher** [comment]
What does your tsconfig look like? I'm assuming that you're not using the `strict` / `strictNullChecks` compiler option?

This rule will not work well without that compiler option.

**Comment by MiK546** [comment]
Oh yeah, that was the problem. I'm still learning TypeScript so I didn't know that by default it includes undefined/null in other types. Using `strictNullChecks` makes everything work as expected. Thanks for the help!
